From 2e3414660cb0c6a024661638d0b237d88b5a7cbc Mon Sep 17 00:00:00 2001
From: Luca Boccassi <bluca@debian.org>
Date: Fri, 22 Dec 2023 01:57:39 +0100
Subject: [PATCH] varlink: avoid logging content of message if it contains
 sensitive data

This is important now that creds are sent via varlink

 systemd-creds[463]: varlink-3: Sending message: {"parameters":{"data":"Zm9vYmFyCg=="}}
 systemd-creds[462]: varlink-3: New incoming message: {"method":"io.systemd.Credentials.Encrypt","parameters":{"data":"Zm9vYmFyCg=="}}

Index: systemd/src/shared/varlink.c
===================================================================
--- systemd.orig/src/shared/varlink.c
+++ systemd/src/shared/varlink.c
@@ -957,10 +957,6 @@ static int varlink_parse_message(Varlink
 
         sz = e - begin + 1;
 
-        varlink_log(v, "New incoming message: %s", begin); /* FIXME: should we output the whole message here before validation?
-                                                            * This may produce a non-printable journal entry if the message
-                                                            * is invalid. We may also expose privileged information. */
-
         r = json_parse(begin, 0, &v->current, NULL, NULL);
         if (r < 0) {
                 /* If we encounter a parse failure flush all data. We cannot possibly recover from this,
@@ -1727,12 +1723,17 @@ Varlink* varlink_flush_close_unref(Varli
 
 static int varlink_format_json(Varlink *v, JsonVariant *m) {
         _cleanup_(erase_and_freep) char *text = NULL;
+        bool sensitive = false;
         int r;
 
         assert(v);
         assert(m);
 
-        r = json_variant_format(m, 0, &text);
+        r = json_variant_format(m, JSON_FORMAT_REFUSE_SENSITIVE, &text);
+        if (r == -EPERM) {
+                sensitive = true;
+                r = json_variant_format(m, /* flags= */ 0, &text);
+        }
         if (r < 0)
                 return r;
         assert(text[r] == '\0');
@@ -1740,7 +1741,7 @@ static int varlink_format_json(Varlink *
         if (v->output_buffer_size + r + 1 > VARLINK_BUFFER_MAX)
                 return -ENOBUFS;
 
-        varlink_log(v, "Sending message: %s", text);
+        varlink_log(v, "Sending message: %s", sensitive ? "<sensitive data>" : text);
 
         if (v->output_buffer_size == 0) {
 
@@ -1771,7 +1772,7 @@ static int varlink_format_json(Varlink *
                 v->output_buffer_index = 0;
         }
 
-        if (json_variant_is_sensitive(m))
+        if (sensitive)
                 v->output_buffer_sensitive = true; /* Propagate sensitive flag */
         else
                 text = mfree(text); /* No point in the erase_and_free() destructor declared above */
Index: systemd/src/shared/json.c
===================================================================
--- systemd.orig/src/shared/json.c
+++ systemd/src/shared/json.c
@@ -1788,2 +1788,5 @@ int json_variant_format(JsonVariant *v,
 
+        if (FLAGS_SET(flags, JSON_FORMAT_REFUSE_SENSITIVE) && json_variant_is_sensitive(v))
+                return -EPERM;
+
         f = memstream_init(&m);
Index: systemd/src/shared/json.h
===================================================================
--- systemd.orig/src/shared/json.h
+++ systemd/src/shared/json.h
@@ -196,6 +196,7 @@ typedef enum JsonFormatFlags {
         JSON_FORMAT_FLUSH       = 1 << 8, /* call fflush() after dumping JSON */
         JSON_FORMAT_EMPTY_ARRAY = 1 << 9, /* output "[]" for empty input */
         JSON_FORMAT_OFF         = 1 << 10, /* make json_variant_format() fail with -ENOEXEC */
+        JSON_FORMAT_REFUSE_SENSITIVE = 1 << 11, /* refuse to format sensitive variants */
 } JsonFormatFlags;
 
 int json_variant_format(JsonVariant *v, JsonFormatFlags flags, char **ret);
